### PR TITLE
fix(general): fix `weekOfSeason` calculation and related stuff (fix #384)

### PR DESCRIPTION
**Pull request by tukusejssirs**
Changes:

- always use `Date.prototype.getUTC*()` methods instead of `Date.prototype.get*()`:
   - in `lib/`, in `tests/`, in `docs/` and in `examples/`;
- fix and improve `getWeekNumber()` function in `lib/utils/dates.ts`;
- fix `If first day of the year 2012 is a Sunday, Mary Mother of God is on that Sunday and the Sunday proceeding will be Epiphany of the Lord` test in `tests/dates.test.ts` (L563):
   - week number is not the same as number of Sundays in a range, thus:
      - Epiphany Sunday should be in the first ISO week of the year;
      - Epiphany should be the 2nd Sunday day of the year:
         - because of this, I had to add `getNumberOfWeekDaysInRange()` function to `lib/utils/dates.ts`, however, I don’t really think we should test how many week days (like Sundays in here) can be found in a range;
         - I think we should remove this test and `getNumberOfWeekDaysInRange()` function.

### CODE REVIEW

**Comment by tukusejssirs** [comment]
@emagnier, I have no idea why [`volta-cli/action@v1`](https://github.com/romcal/romcal/actions/runs/3272042276/jobs/5382590181#step:3:11) would fail. Is zloirock/core-js#648 (closed as `can't reproduce`) related?

Now this PR should fix #386 (without adding tests), thus it is somewhat important to merge ASAP. When you find some time for romcal, you could do this as the first thing.

In regards to #386, I am inclined to update the issue description to provide better description of the bug. Also I want to move the addition/creation of tests into separate issue, where we could first discuss the general structure of the tests (in order to let them be more consistent and more thorough/robust; see #398). WDYT?

Also we should consider adding some tests with timezone(s) different from local one (like `TZ='America/Bogota' npm run test`). Now it all passes (with and without defining a custom timezone), however, we should do it programmatically via tests.

File: lib/utils/dates.ts
Comment (emagnier) [inline lib/utils/dates.ts]: Since this function is only used in tests, not in the main romcal library, could you move it at the top of `dates.test.ts`?

File: lib/utils/dates.ts
Comment (emagnier) [inline lib/utils/dates.ts:70]: 👍

**Review by emagnier — COMMENTED** [review]
Thank you for taking the time to fix this!

There is only one minor comment to address (move `getNumberOfWeekDaysInRange` outside the main romcal library). Other than that, everything LGTM 👍

**Comment by emagnier** [comment]
FYI, I just re-run the failing job and the CI seems happy now (we got the green light).

> In regards to https://github.com/romcal/romcal/issues/386, I am inclined to update the issue description to provide better description of the bug. Also I want to move the addition/creation of tests into separate issue, where we could first discuss the general structure of the tests (in order to let them be more consistent and more thorough/robust; see https://github.com/romcal/romcal/issues/398). WDYT?

I agree 👍  (yes, update the issue according to your comment here)

In addition, we should use Jest snapshots (or something else), to compare data of full generated years (a set of different years with a set of different calendars).

> Also we should consider adding some tests with timezone(s) different from local one (like TZ='America/Bogota' npm run test). Now it all passes (with and without defining a custom timezone), however, we should do it programmatically via tests.

I agree 👍

File: lib/utils/dates.ts
Comment (tukusejssirs) [reply lib/utils/dates.ts]: You’re right, I move it into `dates.test.ts`.

Note, however, that this function is sourced from [SO](https://stackoverflow.com/a/37408412/3408342). I think we could add a source to the JSDoc (which should be changed to TSDoc) and probably merge all those variables, as they are used only once. Something like this:

```ts
/**
 * Gets the number of week days in range, including the end date
 * 
 * @remarks This function was sourced from [StakOverflow](https://stackoverflow.com/a/37408412/3408342).
 *
 * @param    start     - Start date
 * @param    end       - End date
 * @param    dayOfWeek - Day of week to count
 * @returns  Number of week days in the range, including the end date
 */
const getNumberOfWeekDaysInRange = ({start, end, dayOfWeek = 0}: {start: Date, end: Date, dayOfWeek?: 0 | 1 | 2 | 3 | 4 | 5 | 6}): number => {
  // Calculate the number of weeks (even partial) from the first target day to the end
  return Math.ceil(
    // Calculate the number of days from the first target day to the end
    Math.max(
      // Calculate the number of days between start and end, including the end date (thus `+ 1` at the end)
      Math.ceil((end.getTime() - start.getTime()) / 864e5) + 1
      // Calculate the nb of days before the next target day (e.g. next Sunday after start)
      - (7 + dayOfWeek - start.getUTCDay()) % 7,
      0
    ) / 7
  );
}
```

File: lib/utils/dates.ts
Comment (tukusejssirs) [reply lib/utils/dates.ts:70]: Well, I am not that good that algorithm. I sourced it from [SO](https://stackoverflow.com/a/6117889/3408342). :wink:

File: lib/utils/dates.ts
Comment (tukusejssirs) [reply lib/utils/dates.ts]: This should be fixed now.

**Review by emagnier — DISMISSED** [review]
Thank you ❤️ 

Just a small comment, if you modify the PR from a review, it's not worth to squash all the commits together, because it makes the code review more difficult to follow, especially when a lot of code is modified (to see what was adjusted following a comment).
But at the end indeed, it's important to squash everything's with a clean message before merging, for a clean git history (if it's not useful to keep all the commits separated).

**Comment by tukusejssirs** [comment]
> Just a small comment, if you modify the PR from a review, it's not worth to squash all the commits together, because it makes the code review more difficult to follow, especially when a lot of code is modified (to see what was adjusted following a comment). But at the end indeed, it's important to squash everything's with a clean message before merging, for a clean git history (if it's not useful to keep all the commits separated).

Yeah, you’re right. I squashed the changes automatically. :wink:

On the other hand, GitHub provides us a simple [Compare](https://github.com/romcal/romcal/compare/4744345039972cea49a56bc9859570f37a76e139..ab70263668d0d371b9d63b79d49e87961aa1a178) button right below the log about force-pushed commit (right above your latest comment). :wink:

**Comment by emagnier** [comment]
I hadn't noticed this _Compare_ button. A feature that appeared in GitHub without me realizing it 😅. Thank you, it's also a handy tool. So about commits, keep doing as you want 👍

**Comment by tukusejssirs** [comment]
@emagnier, I had to resolve merge conflicts; could you re-approve the PR please? :pray:

**Review by emagnier — APPROVED** [review]

**Comment by tukusejssirs** [comment]
I’ve just noticed that I fixed #384 (which is a discussion) instead of #386, thus is is not automatically close. :man_facepalming:

**Pull request merged**
